### ISSUE TITLE
Broadcast batch elements in vmap

### ISSUE THREAD

**Issue by hamzamerzic** [google/jax#3650]
It would be great to have a way of obtaining elements of a mini batch from within `vmap`, i.e. an equivalent of `pmap's` [all_gather](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.all_gather.html).

Abstracting away the batch dimension seems like one of the core JAX features, but there are a lot of use cases where batch information is necessary. Some that come to mind are normalizing modules, such as BatchNorm, or contrastive estimation which often depends on positive/negative samples across the minibatch.

Related to this, are there any plans to unify `vmap` and `pmap` and potentially expose the rest of the parallel operators for use with `vmap`?

**Comment by mattjj** [comment]
Great questions! Yes indeed there are plans to unify `vmap` and `pmap`, including supporting collectives in `vmap`. Actually, we've _planned_ to do that for a while, but what's new is we've actually made some progress in the last week!

Here's the status. There's a core rewrite branch called "omnistaging3" that moves around some machinery to make this possible (among other things). In addition, we have some experiments in the "gmap" branch, including a proof-of-concept that we can indeed add collective support to `vmap`, as well as some experiments with a map primitive where it's easy to re-schedule the evaluation strategy (between mixtures of parallel, vectorized, and iterative evaluation).

Hopefully we'll have some more updates in the next few weeks!

cc @apaszke @hawkinsp

**Comment by mattjj** [comment]
Actually, I think this might be a duplicate of #2939, so let's close it and coalesce the conversation there!
